File: Algo/Leetcode/243PalindromLinkedList.cpp
// O(n) space. Use stack
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> s;
        ListNode* move = head;
        while(move) {s.push(move->val); move = move->next;}

        move = head;
        while(!s.empty()){
            if(s.top() != move->val) return false;
            else{
                s.pop();
                move = move->next;
            }
        }
        return true;
    }
};


// O(1) space with two pointer. reverse later half list
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head) return true;

        ListNode* dump = new ListNode(0);
        dump -> next = head;
        ListNode* slow = dump;
        ListNode* fast = dump;

        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }

        ListNode* tmp = slow->next;
        slow->next = NULL;
        slow = head;

        tmp = reverseLL(tmp);
        fast = tmp;

        while(slow && fast){
            if (slow->val != fast->val) return false;
            else{
                slow = slow->next;
                fast = fast->next;
            }
        }
        return true;
    }

    ListNode* reverseLL(ListNode* head){
        if(!head) return NULL;

        ListNode* cur = head;
        ListNode* pre = NULL;
        ListNode* tmp = NULL;
        while(cur){
            tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }
        return pre;
    }
};

File: Algo/Leetcode/243PalindromLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        res = []
        while head:
            res.append(head.val)
            head = head.next

        return res == res[::-1]

# O(1) space
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        f = ListNode()
        f.next = head
        fast = f
        slow = f
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        tmp = slow.next
        slow.next = None
        slow = reverseL(tmp)
        fast = head

        while fast and slow:
            if fast.val != slow.val:
                return False
            fast = fast.next
            slow = slow.next

        return True


def reverseL(head):
    tail = head
    if head is None:
        return head

    while tail.next:
        tmp = tail.next
        tail.next = tail.next.next
        tmp.next = head
        head = tmp

    return head